File: MATLAB/FOTO.m
%------------------------------------------------------------------------
% NAME: FOTO.m
% PROJECT: canopy structure project
% PURPOSE: classify image data based upon r-specta
% CATEGORY:
% CALLING SEQUENCE: FOTO(imagefile,windowsize)
% INPUTS: an image layer, a block window size
% PARAMETERS:
% OUTPUTS: rspectra of image blocks
% NOTES: included rank normalization of the image
%        to deal with hazy clouds and uneven illumination
%	 depends on rspectra.m (should be in the matlab path)
% CREATION DATE: 01/07/2013
% AUTHOR: Koen Hufkens, koen.hufkens@gmail.com
% UPDATE: 20/08/2013
%------------------------------------------------------------------------

function rspectra = FOTO(imagefile,windowsize)

	% define global variables for the increment vector
	% and the final output matrix (with rspectra)
	global i output;

	% set counter
	i = 0;

	% convert image to double and average
	% image bands if necessary
	img = imread(imagefile);
	Z = size(img,3);

	if Z == 1;
	    img = double(img);
	else
	    img = double(mean(img,3));
	end

	% get number of blocks that will be processed
	% using the blockproc function below
	[M N] = size(img);
	M = ceil(M/windowsize);
	N = ceil(N/windowsize);
	cells = M*N;

	% output matrix has a maximum vector length of 29
	% only use the first 29 harmonics, if less are
	% produce reduce the output file dimensions
	if windowsize/2 < 29;
	   output = zeros(cells,floor(windowsize/2));
	else
	   output = zeros(cells,29);
	end

	% blockprocess r spectra using rspectrum.m
	% I'm currently working on an octave implementation to go completely GPL
	fun = @(block_struct) rspectrum(block_struct.data,windowsize);
	block_count = blockproc(img,[windowsize windowsize],fun,'PadPartialBlocks',true,'PadMethod','symmetric','TrimBorder',true);

	% finally, write output to csv files
	% reshape the block_count file to get the
	% correct order in the final output file
	% blockproc doesn't calculate blocks perfectly row or column wise
	loc_spectra = reshape(block_count',numel(block_count),1);
	rspectra = output(loc_spectra,:);

	% write rspectra to file, useful when calling
	% the function from a non interactive matlab cli session
	csvwrite('rspectra.csv',output(loc_spectra,:));

	% visualize the rspectra, reduce the dimensionality of the matrix using a
	% principal component analysis, mapping the first 3 PC to RGB colours

	% center the rspectra matrix by subtracting the column mean and dividing by
	% the standard deviation
	norm_rspectra = (rspectra - ones(size(rspectra,1),1)*mean(rspectra)) ./ (ones(size(rspectra,1),1)*var(rspectra).^2);
	norm_rspectra(isnan(norm_rspectra)) = 0; % set NaN to 0, not allowed by princomp
	 
	[~,score,~,~]=princomp(norm_rspectra);
	 
	RGB = zeros(M,N,3);
	RGB(:,:,1) = reshape(score(:,1),M,N);
	RGB(:,:,2) = reshape(score(:,2),M,N);
	RGB(:,:,3) = reshape(score(:,3),M,N);
	 
	% visualize the texture classification
	imshow(RGB);

end 

File: R/FOTO.r
#------------------------------------------------------------------------
# NAME: FOTO.r
# PROJECT: canopy structure project
# PURPOSE: classify image data based upon r-specta
# CATEGORY:
# CALLING SEQUENCE: FOTO(x=rasterlayer,windowsize=61)
# INPUTS: a raster layer, a block window size
# PARAMETERS: plt = plot, TRUE or FALSE
# OUTPUTS: zones (blocks that are processed with the rspectrum.r),
#          output (filled by rspectrum.r, matrix with fft spectra),
#          noutput (normalized output matrix),
#          pcfit (first 3 PC of the noutput file)
#          RGB (rasterbrick containing the first 3 PC for visualization)
# NOTES:
# CREATION DATE: 01/07/2013
# AUTHOR: Koen Hufkens, koen.hufkens@gmail.com
# UPDATE:
#------------------------------------------------------------------------

FOTO <- function(x='',windowsize=61,method="zones",plt=FALSE,rspectrum.n=TRUE){
  
  # 1. load libraries and helper function rspectrum / normalize 
  require(raster)
  #source('rspectrum.r')
  #source('normalize.r')
  
  # standard normalization between 0 and 1
  # this code saves some space further down
  normalize <- function(x){
    (x - min(x, na.rm=TRUE))/(max(x,na.rm=TRUE) - min(x, na.rm=TRUE))
  }
  
  # 2. read the image / matrix file and apply extract the r-spectra
  if (class(x)[1]!="RasterLayer" || class(x)[1]!="RasterBrick" ){  
    # if not a raster object read as image
    img <- brick(x)
    
    # calculate the mean (grayscale)
    # image if there is more than one
    # image band
    if(nbands(img)>1){
      img <- mean(img)
    }
  }
  
  if (class(x)[1]=="RasterLayer"){
    img <- x 
  }
  
  if (class(x)[1]=="RasterBrick"){
    img <- x
    
    # calculate the mean (grayscale)
    # image if there is more than one
    # image band
    if(nbands(img)>1){
      img <- mean(img)
    }
  }
  
  if(method=="zones"|method=="mw"){
    if(method=="zones"){
      # get number of cells to be aggregated to
      N <- ceiling(img@nrows/windowsize)
      M <- ceiling(img@ncols/windowsize)
      cells <- N*M
    }else{
      # get number of cells to be aggregated to
      N <- img@nrows
      M <- img@ncols
      cells <- N*M
    }
  }else{
    print("Choose an available method: zones or mw")
  }
  
  # define output matrix and global increment, i
  if( windowsize/2 < 29){
    output <<- matrix(0,cells,windowsize/2)
  }else{
    output <<- matrix(0,cells,29)
  }
  
  # see rspectrum function above
  i <<- 0
  
  # for every zone execute the r-spectrum function
  if(method=="zones"){
    zones <<- aggregate(img, fact=windowsize, fun=function(x,...){rspectrum(x=x,w=windowsize,n=rspectrum.n,...)},expand=TRUE,na.rm=TRUE)
  }else{
    zones <<- focal(img,matrix(rep(1,windowsize*windowsize),ncol=windowsize,nrow=windowsize), fun=function(x,...){rspectrum(x=x,w=windowsize,n=rspectrum.n,...)},expand=TRUE,na.rm=TRUE)
  }
  
  # 3. reformat the r-spectrum output (normalize) and apply a PCA
  # set all infinite values to NA
  output[is.infinite(output)] <- NA
    
  # normalize matrix column wise (ignoring NA values)
  # (this routine doesn't increase the accuracy of the
  # resulting map, if not it produces worse output. 
  # uncomment the code if wanted)
    
  # normalize column wise by substracting the mean
  # and dividing with the standard deviation
  # keep original input file for reference
  # look up reference!
  noutput <<- (output - apply(output,2,function(output)mean(output,na.rm=T))[col(output)])/apply(output,2,function(output)sd(output,na.rm=T))[col(output)]
    
  # set NA/Inf values to 0 as the pca analys doesn't take NA values
  noutput[is.infinite(noutput) | is.na(noutput)] <- 0
  
  # the principal component analysis
  pcfit <<- princomp(noutput)

  # create reclass files based upon PCA scores
  # only the first 3 PC will be considered
  for (i in 1:3){
    assign(paste('rcl.',i,sep=''),cbind(1:length(pcfit$scores[,i]),normalize(pcfit$scores[,i])),envir=.GlobalEnv)
  }  
  
  # reclassify using the above reclass files
  for (i in 1:3){
    # get() reclass matrix and reclassify the zones file
    assign(paste('PC.',i,sep=''),reclassify(zones,get(paste('rcl.',i,sep=''))),envir=.GlobalEnv)                                
  }
  
  # create a raster brick and plot the RGB image
  # made of pca scores, colours correspond to
  # different scores as split between the scores
  # of the first three pca axis
  img_RGB <- brick(PC.1,PC.2,PC.3)
  
  # 4. plot the classification using an RGB representation of the first 3 PC
  if (plt=="T" | plt=="TRUE"){ # if print is true print the pca classification results
    plot(img,col=gray(0:100/100),legend=FALSE,box=FALSE,axes=FALSE)
    plotRGB(img_RGB,stretch='hist',add=TRUE,alpha=128,bgalpha=0)
    return(img_RGB)
  }

}
